### DISCUSSION TITLE
LLM inference engine

### DISCUSSION THREAD

**Discussion by stvowi** [unleashedpascal/ide#11] [Ideas]
[PasLLM ](https://github.com/BeRo1985/pasllm) is a LLM inference engine in Object Pascal.

This makes it possible to create Unleashed Pascal programmes with AI capabilities without the usual need to install endless dependencies.
I therefore propose that PasLLM be pre-installed in the IDE.

Hey everyone, please give this project as many stars as possible to encourage the developer to keep working on it.

Unfortunately, it’s only a half-finished project.

@see also:

- [**What is LLM inference?**](https://www.ibm.com/think/topics/llm-inference)
- [**PasLLM - AI LLM inference engine in Object Pascal (1)**](https://www.youtube.com/watch?v=xu8bjKOv8b8)
- [**PasLLM - AI LLM inference engine in Object Pascal (2)**](https://www.youtube.com/watch?v=d-kw_9oLTcI)
- [**Lazerus Forum**](https://forum.lazarus.freepascal.org/index.php?topic=72801.0) - PasLLM - LLM Inference Engine in Pure Pascal
- [**Deutschsprachiges Forum zur Entwicklungsumgebung Lazarus**](https://www.lazarusforum.de/viewtopic.php?t=17455&sid=0f1794c0e71dc9387983f56df66c229d) - PasLLM & verschiedene Modelle

The PasLLM project appears to have been put on hold and is not yet ready for production.
It seems one can only make use of what is already available.

**Comment by stvowi** [comment]
You can download PasLLM models from:

- **Pre-converted and quantized models:** [PasLLM model folder on MEGA](https://mega.nz/folder/krcgHCpZ#0tjLqup_Hc4THWC9itDrTg)
- **PasLLM and conversion tools:** [BeRo1985/pasllm on GitHub](https://github.com/BeRo1985/pasllm)
- **Original models:** [Hugging Face](https://huggingface.co/models)

Place downloaded PasLLM models in:

```text
pasllm/bin/models/
```

Example:

```bash
./bin/pasllmcli \
  -model=bin/models/qwen2.5_0.5b_instruct_q40nl.safetensors
```

PasLLM does not directly load arbitrary GGUF files. It uses its own Safetensors-based format with quantization types such as `Q40NL`, `Q40`, `Q80`, `FP8`, and `FP16`. Models from Hugging Face must be converted using `tools/convert.py`, as documented in the [[PasLLM README](https://github.com/BeRo1985/pasllm)](https://github.com/BeRo1985/pasllm).

Good models for initial testing include:

- Qwen 2.5 0.5B Instruct
- Qwen 3 0.6B
- SmolLM2 360M Instruct
- TinyLlama 1.1B Chat

Remember to check the license of each original Hugging Face model.

**Comment by stvowi** [comment]
For Lazarus/Free Pascal development, __the best practical model is Qwen3-Coder 30B_—especially if you want a capable local assistant through Ollama. It is code-specialized, supports a 256K context window, and its MoE design activates only 3.3B parameters per request, making it unusually efficient for its class.  [**ollama - qwen3-coder**](https://ollama.com/library/qwen3-coder) 
For the highest code quality on difficult multi-unit refactors, debugging, and architecture, use a frontier cloud coding model alongside it; but for your 32 GB RAM / 12 GB VRAM local workflow, Qwen3-Coder 30B is the best starting point.
## Recommended setup
Use case | Recommended model | Why
-- | -- | --
Best local coding assistant | qwen3-coder:30b | Purpose-built for agentic software engineering and coding; 19 GB download with 256K context. [ollama]
Larger local/server setup | qwen3-coder-next | Coding-focused 80B MoE model, but the Q4 build alone is 52 GB, so it is not a good fit for your present machine. [ollama]
General Pascal reasoning | qwen3:14b | A useful lighter fallback for explanations, API design, and reviewing compact units. [ollama]
Maximum quality, cloud | Qwen3-Coder 480B Cloud | Stronger for repository-scale work, but the local model requires about 290 GB. [ollama]

## Why this fits Pascal
No leading model is specifically trained or benchmarked primarily on __Free Pascal/Lazarus__, 
so expect occasional Delphi-centric code, incompatible syntax, or assumptions about VCL rather than LCL. The decisive advantage comes from supplying the model with your target details: FPC version, Lazarus/LCL widgetset, operating system, compiler switches, and the involved unit interfaces.
Qwen3-Coder is explicitly positioned for long-context, agentic coding workflows and can be launched through Ollama with coding-agent tooling.  That is particularly useful for your likely tasks: tracing dependencies across units, improving PostgreSQL access code, generating tests, and diagnosing compiler messages. [**ollama - qwen3-coder**](https://ollama.com/library/qwen3-coder)

## Start locally
```
ollama pull qwen3-coder:30b
ollama run qwen3-coder:30b
```
The official Ollama catalogue lists this 30B model at approximately 19 GB and with a 256K context window.  [ollama - qwen3-coder](https://ollama.com/library/qwen3-coder)
If responsiveness or Intel Arc support becomes a limiting factor, use:
```
ollama pull qwen3:14b
ollama run qwen3:14b
```
It will be less reliable on multi-file refactoring, but is a sensible interactive fallback.  [ollama - qwen3:14b ](https://ollama.com/library/qwen3:14b)
## Prompt template
<p>Use a stable project instruction such as:</p><pre><code>You are an expert Free Pascal and Lazarus developer.

Target:
- Free Pascal: 3.2.x
- Lazarus: 3.x
- GUI: LCL, not Delphi VCL
- Platforms: Windows, Linux, macOS
- Database: PostgreSQL via [specify library]
- Compatibility: avoid Delphi-only language and RTL features

Rules:
- Return compilable Object Pascal.
- State required units in the uses clauses.
- Preserve existing public interfaces unless explicitly asked to change them.
- Prefer LCL and FPC-compatible APIs.
- Identify ownership, lifetime, thread-safety, exception handling, and SQL-parameter issues.
- If uncertain about an API, say so instead of inventing it.

Task:
[paste compiler error, unit interface, implementation, and expected behavior]</code></pre><p>For best results, give it a small, coherent slice of the codebase: the failing unit, related interfaces, exact compiler output, and expected behavior. Ask it first to <strong>analyse and propose a patch</strong>, then ask it to emit a complete revised unit; this reduces accidental API changes.</p>

For Lazarus/Free Pascal development, **the best practical model is Qwen3-Coder 30B**—especially if you want a capable local assistant through Ollama. It is code-specialized, supports a 256K context window, and its MoE design activates only 3.3B parameters per request, making it unusually efficient for its class. [[ollama](https://ollama.com/library/qwen3-coder)](https://ollama.com/library/qwen3-coder)

For the *highest* code quality on difficult multi-unit refactors, debugging, and architecture, use a frontier cloud coding model alongside it; but for your 32 GB RAM / 12 GB VRAM local workflow, Qwen3-Coder 30B is the best starting point.

## Recommended setup

| Use case | Recommended model | Why |
|---|---|---|
| Best local coding assistant | `qwen3-coder:30b` | Purpose-built for agentic software engineering and coding; 19 GB download with 256K context.  [ollama](https://ollama.com/library/qwen3-coder)](https://ollama.com/library/qwen3-coder) |
| Larger local/server setup | `qwen3-coder-next` | Coding-focused 80B MoE model, but the Q4 build alone is 52 GB, so it is not a good fit for your present machine.  [ollama](https://ollama.com/library/qwen3-coder-next)](https://ollama.com/library/qwen3-coder-next) |
| General Pascal reasoning | `qwen3:14b` | A useful lighter fallback for explanations, API design, and reviewing compact units.  [[ollama](https://ollama.com/library/qwen3:14b)](https://ollama.com/library/qwen3:14b) |
| Maximum quality, cloud | Qwen3-Coder 480B Cloud | Stronger for repository-scale work, but the local model requires about 290 GB.  [[ollama](https://ollama.com/library/qwen3-coder)](https://ollama.com/library/qwen3-coder) |

## Why this fits Pascal

No leading model is specifically trained or benchmarked primarily on **Free Pascal/Lazarus**, so expect occasional Delphi-centric code, incompatible syntax, or assumptions about VCL rather than LCL. The decisive advantage comes from supplying the model with your target details: FPC version, Lazarus/LCL widgetset, operating system, compiler switches, and the involved unit interfaces.

Qwen3-Coder is explicitly positioned for long-context, agentic coding workflows and can be launched through Ollama with coding-agent tooling.  That is particularly useful for your likely tasks: tracing dependencies across units, improving PostgreSQL access code, generating tests, and diagnosing compiler messages. [[ollama](https://ollama.com/library/qwen3-coder)](https://ollama.com/library/qwen3-coder)

## Start locally

```bash
ollama pull qwen3-coder:30b
ollama run qwen3-coder:30b
```

The official Ollama catalogue lists this 30B model at approximately 19 GB and with a 256K context window. [[ollama](https://ollama.com/library/qwen3-coder)](https://ollama.com/library/qwen3-coder)

If responsiveness or Intel Arc support becomes a limiting factor, use:

```bash
ollama pull qwen3:14b
ollama run qwen3:14b
```

It will be less reliable on multi-file refactoring, but is a sensible interactive fallback. [[ollama](https://ollama.com/library/qwen3:14b)](https://ollama.com/library/qwen3:14b)

## Prompt template

Use a stable project instruction such as:

```text
You are an expert Free Pascal and Lazarus developer.

Target:
- Free Pascal: 3.2.x
- Lazarus: 3.x
- GUI: LCL, not Delphi VCL
- Platforms: Windows, Linux, macOS
- Database: PostgreSQL via [specify library]
- Compatibility: avoid Delphi-only language and RTL features

Rules:
- Return compilable Object Pascal.
- State required units in the uses clauses.
- Preserve existing public interfaces unless explicitly asked to change them.
- Prefer LCL and FPC-compatible APIs.
- Identify ownership, lifetime, thread-safety, exception handling, and SQL-parameter issues.
- If uncertain about an API, say so instead of inventing it.

Task:
[paste compiler error, unit interface, implementation, and expected behavior]
```

For best results, give it a small, coherent slice of the codebase: the failing unit, related interfaces, exact compiler output, and expected behavior. Ask it first to **analyse and propose a patch**, then ask it to emit a complete revised unit; this reduces accidental API changes.

**Reply by stvowi** [reply]
[AI Prompt Engineering Course - Prompt Engineering Beginner COMPLETE Guide and for PROS (2025)](https://www.youtube.com/watch?v=suewaPnOdQI)

**Comment by stvowi** [comment]
Sorry, there is a error in the text, Install Qwen with: ollama run qwen3-coder !

**Comment by stvowi** [comment]
If you want play with PasLLM-Models you can download a [Windows-App](https://mega.nz/folder/krcgHCpZ#0tjLqup_Hc4THWC9itDrTg) .

**Comment by stvowi** [comment]
By the way, the preferred language for prompts is English. Apparently, this gives the best results.

Have fun!

**Comment by stvowi** [comment]
There’s a lovely video on the subject of ‘models’:
[The Best Local AI To Run On 8GB VRAM](https://www.youtube.com/watch?v=DwFZqmVTXB0)

**Comment by stvowi** [comment]
If a model isn’t up to scratch for your purposes, you’ll need to train it using your specialist knowledge.
[Unslot](https://github.com/unslothai/unsloth)  now offers a simple way to improve a model, provided you have the right hardware [hardware ](https://www.youtube.com/watch?v=qhcVT58BByY).

@see also: 
- [First impressions](https://www.youtube.com/watch?v=3ERvw0elyAM)
- [Unsloth's Local Deep Research: Is it any good?](https://www.youtube.com/watch?v=ZdUJDbXF8v4)

**Reply by stvowi** [reply]
A tool as IDE-Addon such like Unslot with/ for pasLLM would be great.

The IDE needs unique features to attract attention and gain popularity.
You can only learn / test Unleashed Pascal with a real project.

**Comment by stvowi** [comment]
Well, have fun with SLM!

@see also:

- [Was ich über lokale KI gelernt habe](https://www.youtube.com/watch?v=M1j_uRqKMKI)
- [This Tiny Model Changes How We Think About Local AI](https://www.youtube.com/watch?v=lsCOkUZxGXw)
- [This tiny LLM dominates RAG and is SUPER FAST](https://www.youtube.com/watch?v=5LTDuOg9DVo)
- [Great Small LLMs for the GPU Poor Like Me!](https://www.youtube.com/watch?v=wJEP4CuR6a4)
- [Cactus Needle - The 26M Function Calling Model](https://www.youtube.com/watch?v=tt9UJ0NiOzU&t=145s)
- [Needle: Finetune a 26M Tool-Calling Model Locally with Ollama](https://www.youtube.com/watch?v=PxgqUFupxUw)
- [Gemma 4 Coder: 12B Model Carrying Fable 5's Reasoning on 8GB VRAM, Fully Offline](https://www.youtube.com/watch?v=7wCdwUgeoNc&t=78s)
- [Introducing Ornith 1.0 - Agentic Coding LLMs](https://www.youtube.com/watch?v=uD4-uy0GmHE)
- [MiniCPM5-2B: The Best Sub-Agent Model Yet?](https://www.youtube.com/watch?v=CXvncR7v66o)

**Comment by stvowi** [comment]
[How to make your local AI 10 times more powerful...](https://www.youtube.com/watch?v=SF_-TbLPpDQ)

**Comment by stvowi** [comment]
[The Best Local AI To Run On 16GB VRAM](https://www.youtube.com/watch?v=-zpmhN5eGuk)

**Comment by stvowi** [comment]
[Frontier-Class AI at Home. Is It Finally Here?](https://www.youtube.com/watch?v=IH8XmxiwliQ)

**Comment by stvowi** [comment]
...that is crazy: [Build the Gemma Translator: No Internet Required](https://www.youtube.com/watch?v=tyELRiuEG40)

**Reply by stvowi** [reply]
[Meet Qwen3.8-LiveTranslate: Qwen's Real-Time Live Interpretation Model](https://www.youtube.com/watch?v=PnXmzT7JAEw)

**Comment by stvowi** [comment]
[The BEST local AI music generator is here!](https://www.youtube.com/watch?v=aUcDyeZgz_k)

**Reply by stvowi** [reply]
[MiniMax 3 Just Killed Suno - Everything You Need To Know](https://www.youtube.com/watch?v=pHcufLW46vg)

**Reply by stvowi** [reply]
[Forget Suno. New BEST local AI music generator is here!](https://www.youtube.com/watch?v=9RtywbN--QE)

**Comment by stvowi** [comment]
[Google Gemma 4 12B: The End of Cloud-Only AI?](https://www.youtube.com/watch?v=vrINTVMbYtk)

**Comment by stvowi** [comment]
Let’s turn Pascal’s source code into an audiobook in a IDE - Addon!? 
Pascal’s syntax is bound to sound better than that of other programming languages.

- [BreezeTTS2 - 100% Local Real-Time Voice](https://www.youtube.com/watch?v=xDHD09fDUkQ)
- [Someone Built Ollama for Audio… and It Runs Locally](https://www.youtube.com/watch?v=CvZuuKkT3Sw)
- [Developers Might Finally Have a Local TTS Model That Doesn’t Suck](https://www.youtube.com/watch?v=pbsTTxKTuts)

**Comment by stvowi** [comment]
[How-To Create Free Local Voice AI with Pipecat + Ollama + Kokoro](https://www.youtube.com/watch?v=iTnpWmty52U)

**Comment by fibodevy** [comment]
If you're looking for people interested in Free Pascal, the official forum is the place to post: https://forum.lazarus.freepascal.org/

That's where the FPC/Lazarus crowd actually is, so you'd get a much bigger response there. It even has a board dedicated to AI in Pascal, which is exactly where PasLLM, local models and tooling recommendations belong. Far more readers and far more replies than anywhere else.

**Reply by stvowi** [reply]
Okay - I respect your wish.
I thought you wanted a few ideas.

Have a good one.

**Reply by fibodevy** [reply]
That's not what I meant at all - ideas are welcome here, keep them coming.

My point was only that you're posting a lot and getting almost no discussion back. The forum has far more people, so a thread there actually turns into a conversation. I read and post there myself, so it's not like I'd miss it.

You can keep posting here, that's fine :smiley: I just don't have much time for watching videos - I've been fixing compiler bugs since this morning. Funnily enough it's AI related: I'm trying to fine-tune a small LLM to write Unleashed Pascal first, and fall back to classic FPC when it can't.

**Reply by stvowi** [reply]
The videos are for your followers, newbies and uneducaded visitors.

**Reply by stvowi** [reply]
I like it that we can fall back to classic FPC. This allows a smooth migration of legacy code to Unleashed Pascal.

**Reply by stvowi** [reply]
Please focus on your roadmap.

**Comment by stvowi** [comment]
[Why You Should Stop Paying For AI (Local AI)](https://www.youtube.com/watch?v=xEI7pvyJ02g)

**Comment by stvowi** [comment]
[LLM Quantization Explained: The Q4 Quality Trap](https://www.youtube.com/watch?v=hYufxef2gvE)

**Comment by stvowi** [comment]
[Python vs Free Pascal (simple speed comparison)](https://www.youtube.com/watch?v=eUG6ub1EEFs)

**Comment by stvowi** [comment]
[The Truth About AGI: Are We Already Past the Point of No Return?](https://www.youtube.com/watch?v=TL3cWILYr6I)
